File: 022 Array of Products/ArrayOfProducts.py
#o(n) time and space

def arrayOfProducts(array):
    # Create a list of ones.
    products = [1 for _ in range(len(array))]

    leftRunningProduct = 1
    # for loop the array length
    for i in range(len(array)):
        products[i] = leftRunningProduct
        leftRunningProduct *= array[i]

    rightRunningProduct = 1
    for i in reversed(range(len(array))):
        products[i] *= rightRunningProduct
        rightRunningProduct *= array[i]

    return products

File: 022 Array of Products/ArrayOfProducts.js
function arrayOfProducts(array) {
  const products = new Array(array.length).fill(1);
	
	let leftRunningProduct = 1;
	for (let i = 0; i < array.length; i++){
		products[i] = leftRunningProduct;
		leftRunningProduct *= array[i]
	}
	let rightRunningProduct = 1;
	for (let i = array.length - 1; i > -1; i--){
		products[i] *= rightRunningProduct;
		rightRunningProduct *= array[i]
	}
	return products;
}
